File: Basic/Test_LuckyNumbersinaMatrix.py
# https://leetcode.com/problems/lucky-numbers-in-a-matrix/

class Solution:
    def luckyNumbers(self, matrix: list):
        # find minimum of each row
        # check if it minimum in its col or not
        result = []
        element = 0
        for i in range(len(matrix)):
            element = min(matrix[i])
            # check
            static = matrix[i].index(element)
            indexUp = i + 0
            indexDown = i + 0
            add = True
            while indexUp >= 0:
                if matrix[indexUp][static] > element:
                    add = False
                    break
                indexUp -= 1
            while indexDown < len(matrix):
                if matrix[indexDown][static] > element:
                    add = False
                    break
                indexDown += 1
            if add:
                result.append(element)
        return result


s = Solution()
print(s.luckyNumbers(matrix=[[3, 7, 8], [9, 11, 13], [15, 16, 17]]))
# 15
print(s.luckyNumbers(matrix=[[1, 10, 4, 2], [9, 3, 8, 7], [15, 16, 17, 12]]))
# 12
print(s.luckyNumbers(matrix=[[7, 8], [1, 2]]))
# 7

File: Basic/Test_LuckyNumbersinaMatrix.c
// https://leetcode.com/problems/lucky-numbers-in-a-matrix/

#include <stdio.h>

int main()
{
    int m, n;
    scanf("%d %d", &m, &n);
    int list[m][n];
    for (int i = 0; i < m; i++)
    {
        for (int j = 0; j < n; j++)
        {
            scanf("%d", &list[i][j]);
        }
    }

    // find the minimum in each row
    // check if it maximum in column or not
    for (int i = 0; i < m; i++)
    {
        // find min in list[i]
        int min = 0;
        int indexUp = 0;
        int staticCol = 0;
        for (int j = 0; j < n; j++)
        {
            if (j == 0)
            {
                min = list[i][j];
                indexUp = i;
                staticCol = j;
            }
            else
            {
                if (list[i][j] < min)
                {
                    min = list[i][j];
                    indexUp = i;
                    staticCol = j;
                }
            }
        }
        int indexDown = indexUp;
        int True = 1;
        while (indexUp >= 0)
        {
            if (list[indexUp][staticCol] > min)
            {
                True = 0;
                break;
            }
            indexUp -= 1;
        }
        while (indexDown < m)
        {
            if (list[indexDown][staticCol] > min)
            {
                True = 0;
                break;
            }
            indexDown += 1;
        }
        if (True == 1)
        {
            printf("%d ", list[i][staticCol]);
        }
    }
    return 0;
}